util/image_viewer/spotfinder_frame.py

```diff
@@ -103,15 +103,13 @@ def __init__(self, *args, **kwds):
                 if not detector:
                     self.params.projection = None
                     continue
-                if len(detector) == 24 and detector[0].get_image_size() == (2463, 195):
-                    self.params.projection = None
-                elif len(detector) == 120 and detector[0].get_image_size() == (
-                    487,
-                    195,
-                ):
+                if detector.has_projection_2d():
                     self.params.projection = None
+                    continue
                 else:
-                    detector.projected_2d = project_2d(detector)
+                    detector.projection_2d_axes = get_detector_projection_2d_axes(
+                        detector
+                    )
```

Comment:
Rather than set an attribute on the detector that is then used later, could we just move the code from `tile_generation.py` that calls `get_panel_projection_2d_from_axes` here and explicitly set the projections in the panel objects. And then the code in `tile_generation.py` can simply look up the projections and use them directly. That way the logic is in a single place, rather than spread over two files.

Reply:
Although I realise the exact projection matrices depend on whether or not the projection is in the lab or image frame. Does this particular point in the code get run again if you've changed the projection mode in the GUI?

Reply:
Had a look at this, and this code does only gets called on startup. I agree it would be cleaner to isolate where projections are set, but it's a bit fiddly in its current state as `get_panel_projection_2d_from_axes` requires image data. I'm also not entirely clear on the benefits of using image projection vs lab. Bit hesitant to extend the surface area of this much further so it might be better tidying up the logic as a separate PR.